### PR TITLE
round robin ip allocation for host-local ipam

### LINKED ISSUE #194: host-local ipam should allocate round-robin ips
host-local ipam tends to allocate ips that are recently used. This sometimes cause unexpected behavior related to arp caching. 

If this sounds like a good idea, I will put together a PR. 

https://github.com/kubernetes/kubernetes/issues/24953

### PR DESCRIPTION
**Pull request by freehan**
plugins: round robin ip allocation for host-local ipam

host-local ipam search for available ips from start of subnet. Hence it tends to allocate ip that are used recently. This is not ideal since it may cause collisions. This PR changes the ip allocation logic to round robin.  

fixes #194
related PR: #196

### CODE REVIEW

**Comment by steveeJ** [comment]
@freehan Thanks for this I'll take a closer look shortly! Please adjust the commit message(s) to conform to https://github.com/containernetworking/cni/blob/master/CONTRIBUTING.md#format-of-the-commit-message, thanks a lot!

@zachgersh @rosenhouse should we request tests written using the Ginkgo/Gomega framework?

**Comment by rosenhouse** [comment]
@freehan We're using [Ginkgo](http://onsi.github.io/ginkgo/) & [Gomega](http://onsi.github.io/gomega/) for writing automated tests.  

You can see examples of this style of testing elsewhere in the repo.  For example, [here](https://github.com/containernetworking/cni/blob/master/pkg/ns/ns_test.go) and [here](https://github.com/containernetworking/cni/blob/master/pkg/types/args_test.go).

We like this style because test descriptions are in plain English and visually close to the test code itself. The Ginkgo test runner is very mature, supporting features like focusing tests and randomly permuting test order.  There are many Gomega matchers that simplify testing of common scenarios (e.g. MatchJSON).

Would you mind re-writing your tests to use these frameworks?

**Comment by freehan** [comment]
Sure. Will resubmit. 

Just wondering if you guys want to have a combination of unit tests and behavior tests. Or only  Ginkgo going forward.

**Comment by rosenhouse** [comment]
@freehan The more tests, the better.  You can write both unit and behavioral tests using Ginkgo & Gomega.

**Comment by freehan** [comment]
Tests Fixed

File: plugins/ipam/host-local/allocator.go [hunk 1]
```diff
@@ -163,3 +163,37 @@ func networkRange(ipnet *net.IPNet) (net.IP, net.IP, error) {
...
+	return ip.NextIP(curIP)
+}
+
+// getSearchRange returns the start and end ip based on the last reserved ip
+func (a *IPAllocator) getSearchRange() (net.IP, net.IP) {
+	var startIP net.IP
+	var endIP net.IP
+	startFromLastReservedIP := false
+	lastReservedIP, err := a.store.LastReservedIP()
+	if err == nil && lastReservedIP != nil {
```
Comment (rosenhouse) [inline plugins/ipam/host-local/allocator.go]: I don't see how this `err` is being handled if it is non-`nil`

File: plugins/ipam/host-local/allocator_test.go [hunk 2]
```diff
@@ -0,0 +1,123 @@
...
+			store := fakestore.NewFakeStore(tc.ipmap, net.ParseIP(tc.lastIP))
+			alloc, _ := NewIPAllocator(&conf, store)
+			res, err := alloc.Get("ID")
+			if tc.expectErr {
+				Expect(err).To(HaveOccurred())
+				continue
+			} else {
+				Expect(err).ToNot(HaveOccurred())
+			}
+			if res.IP.IP.String() != tc.expectResult {
```
Comment (rosenhouse) [inline plugins/ipam/host-local/allocator_test.go]: Why not 
```go
Expect(res.IP.IP.String()).To(Equal(tc.expectResult))
```

File: plugins/ipam/host-local/allocator_test.go [hunk 3]
```diff
@@ -0,0 +1,123 @@
...
+			subnet, err := types.ParseCIDR(tc.subnet)
+			conf := IPAMConfig{
+				Name:   "test",
+				Type:   "host-local",
+				Subnet: types.IPNet{IP: subnet.IP, Mask: subnet.Mask},
+			}
+			store := fakestore.NewFakeStore(tc.ipmap, net.ParseIP(tc.lastIP))
+			alloc, _ := NewIPAllocator(&conf, store)
+			res, err := alloc.Get("ID")
+			if tc.expectErr {
```
Comment (rosenhouse) [inline plugins/ipam/host-local/allocator_test.go]: In canonical BDD style, we don't have assertions inside `if` blocks, but instead separate the different conditions into separate `Context` blocks, e.g.
```
Context("when out of ips", func() {
   BeforeEach(func() {
     // set up the conditions for tests
   })
   It("returns a meaningful error", func() {
      // run the code under test
      Expect(err).To(MatchError("no more ips"))
   })
})
```

**Comment by rosenhouse** [comment]
@freehan Added a few comments on the test coverage but I'm not dogmatic about this.  The fact that you're adding test coverage to a package that previously had none is fantastic.  @steveeJ @containernetworking/cni-maintainers my vote is 👍

File: plugins/ipam/host-local/allocator.go (hunk 1 above)
Comment (freehan) [inline plugins/ipam/host-local/allocator.go]: My intention is to ignore error from store.LastReservedIP, since it is not critical enough to block the whole flow. 

I should still log here. Sounds good?

File: plugins/ipam/host-local/allocator_test.go (hunk 2 above)
Comment (freehan) [inline plugins/ipam/host-local/allocator_test.go]: Fixed

File: plugins/ipam/host-local/allocator_test.go (hunk 3 above)
Comment (freehan) [inline plugins/ipam/host-local/allocator_test.go]: Fixed

**Comment by rosenhouse** [comment]
@freehan You wrote "fixed" below a couple of comments, but I don't see the fixes.  Did you forget to push a commit?

**Comment by freehan** [comment]
I just pushed

**Comment by rosenhouse** [comment]
Great, LGTM
